rest/rest-resources/src/main/java/org/apache/brooklyn/rest/util/MultiSessionAttributeAdapter.java

```diff
@@ -0,0 +1,477 @@
+                
+                // if there is a session marked, use it, unless the server has a preferred session handler and it has an equivalent session
+                // this way if a session is marked (from use in a context where we don't have a web request) it will be used
+                SessionHandler preferredHandlerForMarkedSession = findPeerSessionMarkedPreferred(localSession.getId(), handlers);
+                if (preferredHandlerForMarkedSession!=null) return preferredHandlerForMarkedSession;
+                
+                // nothing marked as preferred; if server global handler has a session, mark it as preferred
+                // this way it will get found quickly on subsequent requests
+                if (sessionAtServerGlobalPreferredHandler!=null) {
+                    sessionAtServerGlobalPreferredHandler.setAttribute(KEY_IS_PREFERRED, true);
+                    return preferredServerGlobalSessionHandler; 
+                }
+                
+                if (allowHandlerThatDoesntHaveSession && preferredServerGlobalSessionHandler!=null) {
+                    return preferredServerGlobalSessionHandler;
+                }
+                if (preferredServerGlobalSessionHandler==null) {
+                    preferredServerGlobalSessionHandler = findPreferredBundleHandler(localSession, server, handlers);
+                    if (preferredServerGlobalSessionHandler!=null) {
+                        // recurse
+                        return getPreferredJettyHandler(localSession, allowHandlerThatDoesntHaveSession, markAndReturnThisIfNoneFound);
+                    }
+                }
+    
+                if (markAndReturnThisIfNoneFound) {
+                    // nothing detected as preferred ... let's mark this session as the preferred one
+                    markSessionAsPreferred(localSession, " (this is the handler that the request came in on)");
+                    return localHandler;               
+                }
+                
+            } else {
+                log.warn("Could not find server for "+info(localSession));
+            }
+            return null;
+        }
+    
+        protected void markSessionAsPreferred(HttpSession localSession, String msg) {
+            if (log.isTraceEnabled()) {
+                log.trace("Recording on "+info(localSession)+" that it is the preferred session"+msg);
+            }
+            localSession.setAttribute(KEY_IS_PREFERRED, true);
+        }
+    
+        protected SessionHandler findPreferredBundleHandler(Session localSession, Server server, Handler[] handlers) {
+            if (PREFERRED_SYMBOLIC_NAME==null) return null;
+            
+            SessionHandler preferredHandler = null;
+            
+            if (handlers != null) {
+                for (Handler handler: handlers) {
+                    SessionHandler sh = (SessionHandler) handler;
+                    ContextHandler.Context ctx = getContext(sh);
```

Comment:
I'm a bit concerned performance wise about this call, as `getContext(sh)` uses refection. This is done within a loop of handler which can be potentially big.

If I read this code correctly, this will be executed only during the session creation, if there is no other sessions marked as preferred, so it is somewhat mitigated. But still, I'm wondering if it could be a bottleneck